app/models/component.rb

```diff
@@ -16,27 +16,35 @@ class Component < Aggregate
              -> { where(usage: CONSTANTS::ENCRYPTION) }, class_name: 'Certificate', optional: true
 
   def self.to_service_metadata(event_id, published_at = Time.now)
-    matching_service_adapters = MsaComponent.includes(:enabled_signing_certificates,
-                                                      :encryption_certificate)
-    service_providers = SpComponent.includes(:enabled_signing_certificates,
-                                             :encryption_certificate)
+    matching_service_adapters = MsaComponent.all_components
+    service_providers = SpComponent.all_components
     components = matching_service_adapters + service_providers
 
     {
       published_at: published_at,
       event_id: event_id,
-      connected_services: components.map(&:services_to_metadata).flatten,
+      connected_services: components.map(&:services_to_metadata).flatten.compact,
       matching_service_adapters: matching_service_adapters.map(&:to_metadata),
       service_providers: service_providers.map(&:to_metadata)
     }
   end
 
+  def self.all_components
+    self.joins(:certificates)
```

Comment:
What's the reason for changing from the below?

```
.includes(:enabled_signing_certificates, :encryption_certificate)
```

Reply:
I should have included the reason in my PR comment.

The `has_many`'s are used when showing components. However because the `certificates` only have a `component_id` attribute then there is a possibility to return certificates attached to both component types.

The next PR after this contains the change to pass a `component` into the `has_many` so that it can only return the services specific to that type. That change means it's no longer possible to use them in the class method.

Reply:
Would giving the Certificate a polymorphic association to a Sp or Msa fix this?

Reply:
Yes indeed, we could make it like the Services, but that would require us to have columns for both the SP and MSA components and we were trying to keep it to using just one component id.

Reply:
I'm not sure I follow what you mean. If you add a `component_type` column to the certificates table then you will be able to go back to what was previously happening and continue using Rails' association helpers. Otherwise, you're going to have to start re-implementing these features and I expect we'll run into quite a few bugs along the way.